org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/IndexManagerTests.java

```diff
@@ -178,6 +192,103 @@ public void testSearchMetaIndex_ForSourceTypeDeclarations() throws CoreException
 		assertEquals("No results found", 1, indexNames.get().size());
 	}
 
+	public void testGH1203() throws Exception {
+		waitUntilIndexesReady();
+
+		addLibrary(this.project, "lib.jar", "lib.zip",
+				new String[] {
+					"test/XY.java",
+					"""
+					package test;
+					public class XY {}
+					"""
+				},
+				"1.8");
+
+		// Wait a little bit to be sure file system is aware of zip file creation
+		try {
+			Thread.sleep(1000);
+		}
+		catch (InterruptedException ie) {
+			// skip
+		}
+
+		String packageFolder = "/" + this.project.getElementName() + "/src/test";
+		createFolder(packageFolder);
+		String source =
+				"""
+				package test;
+				public class ZXX {
+				}
+				""";
+		boolean indexDisabled = this.indexDisabledForTest;
+		this.indexDisabledForTest = true; // avoid waitForIndex() call from ModifyingResourceTests.createFile(String, InputStream)
+		createFile(packageFolder+"/ZXX.java", source);
+		this.indexDisabledForTest = indexDisabled;
+		this.workingCopies = new ICompilationUnit[2];
+		this.workingCopies[0] = getCompilationUnit(packageFolder+"/XX.java");
+
+		IPath prjPath = this.project.getProject().getLocation();
+		String jarPath = prjPath+"/lib.jar";
+		createJar(new String[] {
+			"test/ZXY.java",
+			"package test;\n" +
+			"public class ZXY {}\n",
+			"test/ZXZ.java",
+			"package test;\n" +
+			"public class ZXZ {}\n",
+		}, jarPath);
+		this.project.getProject().refreshLocal(IResource.DEPTH_INFINITE, null);
+
+		try {
+			Thread.sleep(1000);
+		}
+		catch (InterruptedException ie) {
+			// skip
+		}
+
+		removeLibrary(this.project, "lib.jar", "lib.zip");
+
+		Thread.sleep(2000);
+
+		this.workingCopies[1] = getWorkingCopy(packageFolder+"/ZX0.java",
+				"""
+				package test;
+				public class ZX0 {}
+				""");
+		searchTypesExpecting("ZX", new HashSet<>(Arrays.asList("ZXX", "ZX0")));
+	}
+
+	public void searchTypesExpecting(String namePrefix, Set<String> names) throws JavaModelException {
+		IRestrictedAccessTypeRequestor typeRequestor = new IRestrictedAccessTypeRequestor() {
+			@Override
+			public void acceptType(int modifiers, char[] packageName, char[] simpleTypeName, char[][] enclosingTypeNames,
+					String path, AccessRestriction access) {
+				if (!names.remove(String.valueOf(simpleTypeName))) {
+					fail("unexpected type "+String.valueOf(simpleTypeName));
+				}
+			}
+		};
+		try {
+			new BasicSearchEngine(this.workingCopies).searchAllTypeNames(
+				null,
+				SearchPattern.R_EXACT_MATCH,
+				namePrefix.toCharArray(),
+				SearchPattern.R_PREFIX_MATCH,
+				IJavaSearchConstants.TYPE,
+				BasicSearchEngine.createJavaSearchScope(false, new IJavaElement[] {this.project}),
+				false,
+				typeRequestor,
+				IJavaSearchConstants.CANCEL_IF_NOT_READY_TO_SEARCH,
```

Comment:
CANCEL_IF_NOT_READY_TO_SEARCH generates OperationCanceledException if indexer is still working.
you need to use WAIT_UNTIL_READY_TO_SEARCH.
However the test then will run into a deadlock if notifyIdle() requests a new MetaIndexJob. I will propose a fix.

Reply:
> you need to use WAIT_UNTIL_READY_TO_SEARCH.

This implies that I need to go back to the original #1203 to check if the fix in #1283 has improved anything, because that's what the test case was distilled from - use of `CANCEL_IF_NOT_READY_TO_SEARCH` was not my invention.

Reply:
Here's how `CANCEL_IF_NOT_READY_TO_SEARCH` came into the picture, initially:
```
SearchableEnvironment.findTypes(char[], boolean, int, int, boolean, ISearchRequestor, IProgressMonitor) line: 747	
CompletionEngine.findTypesAndPackages(char[], Scope, boolean, boolean, ObjectVector) line: 11933	
CompletionEngine.completionOnFieldType(ASTNode, Scope) line: 2751	
CompletionEngine.complete(ASTNode, ASTNode, ASTNode, CompilationUnitDeclaration, Binding, Scope, boolean) line: 2014	
CompletionEngine.complete(ICompilationUnit, int, int, ITypeRoot) line: 2359	
CompilationUnit(Openable).codeComplete(ICompilationUnit, ICompilationUnit, int, CompletionRequestor, WorkingCopyOwner, ITypeRoot, IProgressMonitor) line: 136	
CompilationUnit.codeComplete(int, CompletionRequestor, WorkingCopyOwner, IProgressMonitor) line: 367	
CompilationUnit.codeComplete(int, CompletionRequestor, WorkingCopyOwner) line: 362	
CompletionTests.testCamelCaseType1() line: 759	
```

Instead of `WAIT_UNTIL_READY_TO_SEARCH` this code location has a catch clause, falling back to `SearchableEnvironment.findTypes(String, ISearchRequestor, int)`, where explicit traversal of package fragment roots is performed instead of search. Perhaps that fallback is inferior/buggy, so occurrence of `OperationCanceledException` should probably be acceptable after all.